src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs

```diff
@@ -0,0 +1,2348 @@
+                .Where(entry => entry.ExpectedJson != null)
+                .Select(entry => new object[] { entry });
+
+        [Fact]
+        public async Task PolymorphicInterface_CustomConfigWithNearestAncestorFallback_TestDataArray_Serialization()
+        {
+            IEnumerable<(PolymorphicInterface Value, string ExpectedJson)> inputs =
+                PolymorphicInterface.Helpers.GetSerializeTestData_CustomConfigWithNearestAncestorFallback()
+                    .Where(entry => entry.ExpectedSerializationException is null)
+                    .Select(entry => (entry.Value, entry.ExpectedJson));
+
+            await TestMultiContextSerialization(inputs, options: PolymorphicInterface.Helpers.CustomConfigWithNearestAncestorFallback);
+        }
+
+        [Fact]
+        public async Task PolymorphicInterface_CustomConfigWithNearestAncestorFallback_TestDataArray_Deserialization()
+        {
+            IEnumerable<(string ExpectedJson, PolymorphicInterface ExpectedRoundtripValue)> inputs =
+                PolymorphicInterface.Helpers.GetSerializeTestData_CustomConfigWithNearestAncestorFallback()
+                .Where(entry => entry.ExpectedRoundtripValue is not null)
+                .Select(entry => (entry.ExpectedJson, entry.ExpectedRoundtripValue));
+
+            await TestMultiContextDeserialization(
+                inputs,
+                options: PolymorphicInterface.Helpers.CustomConfigWithNearestAncestorFallback,
+                equalityComparer: PolymorphicEqualityComparer<PolymorphicInterface>.Instance);
+        }
+
+        // --
+
+        [Theory]
+        [MemberData(nameof(Get_PolymorphicInterface_DiamondInducingConfigurations_ShouldThrowNotSupportedException))]
+        public async Task PolymorphicInterface_DiamondInducingConfigurations_ShouldThrowNotSupportedException(PolymorphicInterface value, JsonPolymorphicTypeConfiguration configuration)
+        {
+            var options = new JsonSerializerOptions { PolymorphicTypeConfigurations = { configuration } };
+            await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.SerializeWrapper(value, options));
+        }
+
+        public static IEnumerable<object[]> Get_PolymorphicInterface_DiamondInducingConfigurations_ShouldThrowNotSupportedException()
+            => PolymorphicInterface.Helpers.GetDiamondInducingConfigurations().Select(entry => new object[] { entry.diamondValue, entry.configuration });
+
+
+        [JsonDerivedType(typeof(DerivedClass_NoTypeDiscriminator))]
+        [JsonDerivedType(typeof(DerivedClass_TypeDiscriminator), "derivedClass")]
+        [JsonDerivedType(typeof(DerivedStruct_NoTypeDiscriminator))]
+        [JsonDerivedType(typeof(DerivedStruct_TypeDiscriminator), "derivedStruct")]
+        [JsonDerivedType(typeof(DerivedInterface1.ImplementingClass), "implementingClassOfDerivedInterface")]
+        public interface PolymorphicInterface
```

Comment:
Preference\style: perhaps add several new .cs files to help break this up instead of placing a bunch of classes inside of an interface for grouping reasons?

Reply:
Nesting types became a bit of necessity due to scoping concerns. There's a large number of polymorphic types that define nest types like `DerivedClass` or `DerivedCollection`, and flattening those would require some kind of disambiguation like `PolymorphicClass_DerivedClass`. Nesting feels like a natural way to address this and saves a few characters as well in the attribute declarations.

Breaking it up to more files might make sense actually, this file is pretty huge now.